pkg/raft/term_cache.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2025 The Cockroach Authors.
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package raft
+
+// TermCache represents a compressed representation of the suffix of a raftLog.
+// It is used to determine the term of a raft entry at a specific index.
+//
+// The TermCache conforms to raft safety properties. It also provides the
+// following guarantees:
+//
+//  1. cache[i].Index < entries[i+1].Index,
+//  2. cache[i].term < entries[i+1].Term.
+//  3. cache[0] is always populated.
+//  4. Except for the first entry cache[0], all entries are term flips.
+//
+// Which we can use the above invariants to determine the term of a
+// raft entry if given an index that is greater than cache[0].index.
+//
+// Users of this struct can assume the invariants hold true.
+type TermCache struct {
+	// cache is a fixed size slice storing term flip entryIDs.
+	// Special case: cache[0] may not be a term flip entry,
+	// but is always populated.
+	cache []entryID
+	// maxSize is the maximum size of the term cache.
+	maxSize uint64
+}
+
+// NewTermCache initializes a TermCache with a fixed size and the first entry.
+func NewTermCache(size uint64, eID entryID) *TermCache {
+	tc := &TermCache{
+		cache:   make([]entryID, 0, size),
+		maxSize: size,
+	}
+	tc.resetWithFirst(eID)
+	return tc
+}
+
+// Term returns the entry term based on the given entry index.
+// Returns error if not in the term cache.
+func (tc *TermCache) Term(index uint64) (term uint64, found bool) {
+	if len(tc.cache) == 0 {
+		panic("term cache should not be empty when calling term")
+	}
+
+	for i := len(tc.cache) - 1; i >= 0; i-- {
+		if index >= tc.cache[i].index {
+			return tc.cache[i].term, true
+		}
+	}
+
+	return 0, false
+}
+
+// ScanAppend appends a slice of raftLog to the TermCache.
+// If overwriting, it will truncate the cache first
+// and append the new entries.
+func (tc *TermCache) ScanAppend(entriesSlice LogSlice) {
+	entries := entriesSlice.Entries()
+
+	if len(tc.cache) == 0 {
+		panic("Term cache should not be empty when appending entries in batch")
+	}
+
+	tc.truncateFrom(entriesSlice.prev)
+
+	for _, eID := range entries {
```

Comment:
nit: can just inline the `Entries()` call
nit: `eID` doesn't reflect the type (which is `pb.Entry`), needs a rename

```suggestion
	for _, entry := range ls.Entries() {
```